### DISCUSSION TITLE
sdcard.py and os.chdir() smash subsequent import

### DISCUSSION THREAD

**Discussion by smhodge42** [micropython/micropython#15472] [ESP32]
Espressif ESP32-S3 on DevKitC-1, latest version of uPython 1.23. Using the <sdcard.py> module for access to a uSD card.

The following REPL session, which is extracted from my main application, works fine:
```
import sdcard
import os
from machine import Pin, SPI
spi = SPI(1, sck=11, mosi=10, miso=9)
cs = Pin(3, mode=Pin.OUT)
sd = sdcard.SDCard(spi,cs)
vfs = os.VfsFat(sd)
os.mount(vfs,"/log/")
os.listdir('/log/')
from psr import psr
```
Files get listed and the last line imports successfully from a module already on the ESP (used in the main application).  However, if I use os.chdir() to change the CWD to the mount point, the files still list fine but the import does not:

```
import sdcard
import os
from machine import Pin, SPI
spi = SPI(1, sck=11, mosi=10, miso=9)
cs = Pin(3, mode=Pin.OUT)
sd = sdcard.SDCard(spi,cs)
vfs = os.VfsFat(sd)
os.mount(vfs,"/log/")
os.chdir('/log/')
os.listdir()
from psr import psr
>>> ImportError: no module named 'psr'
```

Obviously, I know what the work-around is, don't chdir(). I'm posting this in case it's a bug or if someone can tell me what I'm missing here. Thanks

**Comment by shariltumin** [comment]
This is because the current working directory (CWD) does not contain the file you are trying to import.

```
>>> import sys
>>> sys.path
['', '.frozen', '/lib']
```

The ```import``` will look for a module file in CWD (''), all frozen modules ('.frozen') all mip.install modules (/lib').

If you did ```os.chdir('/log')``` your CWD is now '/log̈́' and the '' of sys.path refers to all files in '/log'. I do not know where you put your "psr.py" module and if it depends on other modules.

If the "psr" depends only on modules in ```sys.path``` listed above, then you can try:

```python
psr = __import__('/psr').psr # Assuming /psr.py exists!
```

Hope this helps.

**Reply by mendenm** [reply]
wouldn't it work just to add the appropriate absolute root to sys.path?

**Reply by shariltumin** [reply]
I am using ```mpremote```
```python
import sys
sys.path = ['', '.frozen', '/', '/lib', '/remote']
```
should work.

The '' should take care of all module files in the current working directory.

**Comment by sosi-deadeye** [comment]
Never change the current working directory inside a program. Instead, you should use paths.

```python
import sdcard
import os

from machine import Pin, SPI


LOGS = "/log"

spi = SPI(1, sck=11, mosi=10, miso=9)
cs = Pin(3, mode=Pin.OUT)
sd = sdcard.SDCard(spi, cs)
vfs = os.VfsFat(sd)

os.mount(vfs, LOGS)
os.listdir(LOGS)


# usually imports are on the top
from psr import psr

```

**Reply by mendenm** [reply]
OK, why not change the working directory inside code? That's what `os.chdir()` is there for.  It can greatly simplify file management and readability of code.    It seems more appropriate to  just make sure your module path is in sys.path, before fiddling.  `sys.path.append(os.path.abspath('.'))` near the top of the file is likely to fix most issues.

**Reply by robert-hh** [reply]
There is not os.path in MicroPython.

**Comment by smhodge42** [comment]
I am unable to open files that way. That’s why I posted.

f = open(/log/name.ext) does not work, comes back with a ‘no entity” (IIRC). The only way I can open files on the SD card is to change the CWD to ‘/log/’ and then f=open(name.ext) works just fine.

I got my rather big and complicated ESP32 program working just fine by using doing the CWD and appending ‘/’ to sys.path, which fixes the import issue.

os.chdir(‘/log’)
sys.path.append(‘/’)

**Comment by mendenm** [comment]
Oops, good point.  Can't use `abspath()`.

**Comment by smhodge42** [comment]
We seem to be going in circles here.  That seems to be exactly what I did to get it working, as I posted yesterday:

os.chdir('/log/')
sys.path.append('/')

The default system path was ['', '.frozen', '/lib']. It is now ['', '.frozen', '/lib', '/']

I can open files and do imports. Issue solved as far as I am concerned.
